### ISSUE TITLE
provider default labels create issue on PSC forwarding rule create

### ISSUE THREAD

**Issue by Maarc-D** [hashicorp/terraform-provider-google#17334]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to a user, that user is claiming responsibility for the issue.
* Customers working with a Google Technical Account Manager or Customer Engineer can ask them to [reach out internally](https://github.com/hashicorp/terraform-provider-google/wiki/Customer-Contact#raising-gcp-internal-issues-with-the-provider-development-team)
  to expedite investigation and resolution of this issue.

### Terraform Version

1.7.3

### Affected Resource(s)

google_compute_forwarding_rule

### Terraform Configuration

```tf
resource "google_compute_forwarding_rule" "this" {
  for_each                = local.gsn-psc-filtered-services-attachements
  name                    = each.value.name
  region                  = var.subnetwork.region
  project                 = var.project
  load_balancing_scheme   = "" # need to override EXTERNAL default when target is a service attachment
  network                 = var.subnetwork.network
  allow_psc_global_access = true
  subnetwork              = var.subnetwork.id
  target                  = each.value.id
  ip_address              = google_compute_address.psc_ilb_consumer_address[each.value.url].id
}
```

### Debug Output

![image](https://github.com/hashicorp/terraform-provider-google/assets/62051944/a6a4d33e-6c1a-4deb-8d52-44f266137f06)

### Expected Behavior

work fine

### Actual Behavior

![image](https://github.com/hashicorp/terraform-provider-google/assets/62051944/2b4494a6-2f41-45b6-8ef6-bef8d6c3a0d6)

### Steps to reproduce

1. `terraform apply`

### Important Factoids

if I remove default label on provider it create and If I reposition them it update successfully ;)

**Comment by melinath** [comment]
Marking as service/terraform for now since this seems to be related to default provider labels. @rileykarson @zli82016 FYI

**Comment by zli82016** [comment]
@Maarc-D , can you please provide the configuration for provider default labels?

**Comment by Maarc-D** [comment]
```hcl
provider "google" {
  project        = local.project_id
  region         = var.region
  default_labels = local.global_labels
}
```

**Comment by AndreasBergmeier6176** [comment]
This seems to be related to https://github.com/hashicorp/terraform-provider-google/issues/16564, which was perhaps prematurely locked.

**Comment by zli82016** [comment]
> ```terraform
> google_compute_forwarding_rule
> ```

Thanks for providing the configuration. What are values of local.global_labels?

Also, is it possible to provide more logs by running the command `TF_LOG=DEBUG terraform apply`? Thanks.

**Comment by AndreasBergmeier6176** [comment]
> > ```terraform
> > google_compute_forwarding_rule
> > ```
> 
> Thanks for providing the configuration. What are the values of local.global_labels?

It does not matter. As soon as there are default labels `google_compute_forwarding_rule` fails on creation.
So say:

```tf
locals {
  global_labels = {"foo": "bar"}
}
```

**Comment by AndreasBergmeier6176** [comment]
At this point IMO the practical fix would be for the Provider to implement `create` as:
1. create Resource on GCP - ignoring any labels
1. apply labels afterwards

Then the shortcoming of the GCP Backend gets at least handled somehow.

**Comment by zli82016** [comment]
> 
> It does not matter. As soon as there are default labels `google_compute_forwarding_rule` fails on creation. So say:
> 
> ```terraform
> locals {
>   global_labels = {"foo": "bar"}
> }
> ```

Sorry, I still cannot reproduce this issue. The resource creation succeeded when I tried to create the resource with provider default labels.

Can you please provide more logs by running the command TF_LOG=DEBUG terraform apply? Thanks.

**Comment by rileykarson** [comment]
@zli82016 can you share your configuration? This will only fail on one of the more involved variants of the resource, a PSC forwarding rule, and https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_forwarding_rule#example-usage---forwarding-rule-vpc-psc is a good baseline configuration.

**Comment by zli82016** [comment]
My configuration is
```
provider "google" {
  default_labels = {
    default_key = "test"      
  }     
}

resource "google_compute_target_pool" "foobar-tp" {
  description = "Resource created for Terraform acceptance testing"
  instances   = ["us-central1-a/foo", "us-central1-b/bar"]
  name        = "tf-test-1"
}

resource "google_compute_forwarding_rule" "foobar" {
  description = "Resource created for Terraform acceptance testing"
  ip_protocol = "TCP"
  name        = "tf-test-1"

 load_balancing_scheme = "INTERNAL"
  backend_service       = google_compute_region_backend_service.producer_service_backend.id
  network_tier = "PREMIUM"
 all_ports             = true
}

resource "google_compute_region_backend_service" "producer_service_backend" {
  name     = "tf-test-producer-service-backend"
  region   = "us-central1"

  health_checks = [google_compute_health_check.producer_service_health_check.id]
}

resource "google_compute_health_check" "producer_service_health_check" {
  name     = "tf-test-producer-service-health-check"

  check_interval_sec = 1
  timeout_sec        = 1
  tcp_health_check {
    port = "80"
  }
}

```

**Comment by zli82016** [comment]
> @zli82016 can you share your configuration? This will only fail on one of the more involved variants of the resource, a PSC forwarding rule, and https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_forwarding_rule#example-usage---forwarding-rule-vpc-psc is a good baseline configuration.

I can reproduce it with the PSC forwarding rule.

**Comment by petemounce** [comment]
Confirmed reproduced with terraform `1.8.5` & hashicorp/google `5.35.0`.

Also related; #16255, perhaps also closed in error. Specifically this comment https://github.com/hashicorp/terraform-provider-google/issues/16255#issuecomment-1773031396.

**Comment by RomainBre** [comment]
Any news ? Facing this issue with hashicorp/google v5.37.0...

**Comment by zli82016** [comment]
The root cause of the issue is not provider default labels. It looks like the `labels` field should not be included in the CREATE request when creating PSC forwarding rule.
https://github.com/hashicorp/terraform-provider-google/blob/main/google/services/compute/resource_compute_forwarding_rule.go#L722-L727

After the resource is created, there is a separate request to set labels. 
https://github.com/hashicorp/terraform-provider-google/blob/main/google/services/compute/resource_compute_forwarding_rule.go#L806

1. Should the issue be forwarded to the service team?
2. The pre-create code can be added to remove `labels` from CREATE request object. Should it be added to PSC forwarding rules or all of forwarding rules?

**Comment by rileykarson** [comment]
> The root cause of the issue is not provider default labels. It looks like the labels field should not be included in the CREATE request when creating PSC forwarding rule.

I think default labels exacerbate the issue, though!

1. Yes. We may want to resolve this ourselves since it's a complicated change, though.
2. Just PSC I think-  we want to add values during the initial creation whenever possible. Labels are used to tag billing information, and customers could end up with marginal amounts of their bill missing a cost center between creation and adding the label.

**Comment by Maarc-D** [comment]
still got this with 5.40.0
```
- Using previously-installed hashicorp/google v5.40.0
```
```
│ Error: Error creating ForwardingRule: googleapi: Error 400: Invalid value for field 'resource.labels': ''. Invalid field set in Private Service Connect Forwarding Rule. This field should not be set., invalid
│ 
│   with module.gsn-consumer.google_compute_forwarding_rule.this["itdiscovery.global.worldline-solutions.com"],
│   on .terraform/modules/gsn-consumer/main.tf line 40, in resource "google_compute_forwarding_rule" "this":
│   40: resource "google_compute_forwarding_rule" "this" {
```

**Comment by zli82016** [comment]
> still got this with 5.40.0
> 
> ```
> - Using previously-installed hashicorp/google v5.40.0
> ```
> 
> ```
> │ Error: Error creating ForwardingRule: googleapi: Error 400: Invalid value for field 'resource.labels': ''. Invalid field set in Private Service Connect Forwarding Rule. This field should not be set., invalid
> │ 
> │   with module.gsn-consumer.google_compute_forwarding_rule.this["itdiscovery.global.worldline-solutions.com"],
> │   on .terraform/modules/gsn-consumer/main.tf line 40, in resource "google_compute_forwarding_rule" "this":
> │   40: resource "google_compute_forwarding_rule" "this" {
> ```

The fix will be in 5.41.0
